Analyses/src/BFieldPlotter_module.cc

```diff
@@ -0,0 +1,173 @@
+					<< _plane.c_str();
+    }
+
+  } //end constructor
+
+  void BFieldPlotter::analyze(const art::Event& event) {
+    art::ServiceHandle<art::TFileService> tfs;
+    // Get the magnetic field manager and the map lists
+    GeomHandle<BFieldManager> bf;
+    const BFieldManager::MapContainerType innerMaps = bf->getInnerMaps();
+    const BFieldManager::MapContainerType outerMaps = bf->getOuterMaps();
+ 
+    //plot inner maps
+    for(std::shared_ptr<BFMap> map : innerMaps) { 
+      fillHistogram(map, tfs);
+    } 
+    
+    //plot outer maps
+    for(std::shared_ptr<BFMap> map : outerMaps) { 
+      fillHistogram(map, tfs);
+    } 
+
+    //plot the default field returned from the manager
+    fillHistogram(NULL, tfs);
+  } // end analyze
+
+
+  //fill a histogram with the magnetic field
+  void BFieldPlotter::fillHistogram(std::shared_ptr<BFMap> map, art::ServiceHandle<art::TFileService> tfs) {
+    //define histogram binning
+    int nbinsOne = (_axisOneMax - _axisOneMin)/_mapBinSize + 1;
+    int nbinsTwo = (_axisTwoMax - _axisTwoMin)/_mapBinSize + 1;
+    
+    //if a null map, plot the default field from the manager
+    const std::string name = (map) ? map->getKey() : "default";
+
+    if(_hMap[name]) return; //already filled the maps in a previous event
+
+    //make a directory for the map
+    art::TFileDirectory tfdir = tfs->mkdir( ("BFieldMapper_"+name).c_str() );
+
+    //define a histogram
+    _hMap[name] = tfdir.make<TH2F>(("hMap"+name).c_str()  , (name + " Magnetic field map").c_str(), 
+				   // add offsets so all values are bin centers and fit edge into the map
+				   nbinsOne, _axisOneMin - _mapBinSize/2.,_axisOneMax + _mapBinSize/2.,
+				   nbinsTwo, _axisTwoMin - _mapBinSize/2.,_axisTwoMax + _mapBinSize/2.);
+    
+    GeomHandle<BFieldManager> bf; //only needed in null map case
+    //Loop through the points in the magnetic field
+    double axisOne = _axisOneMin;
+    while(axisOne <= _axisOneMax) {
+      double axisTwo = _axisTwoMin;
+      while(axisTwo <= _axisTwoMax) {
+	CLHEP::Hep3Vector point;
+	CLHEP::Hep3Vector field;
+	if(_plane == "x") 
+	  point = CLHEP::Hep3Vector(_planeValue, axisOne, axisTwo);
+	else if(_plane == "y") 
+	  point = CLHEP::Hep3Vector(axisOne, _planeValue, axisTwo);
+	else if(_plane == "z") 
+	  point = CLHEP::Hep3Vector(axisOne, axisTwo, _planeValue);
+	if(map)
+	  map->getBFieldWithStatus(point,field);
+	else //if null, get the manager and plot the field returned by it
+	  bf->getBFieldWithStatus(point,field);
+
+	_hMap[name]->Fill(axisOne, axisTwo, field.mag()); //fill with weight of the field magnitude
+	axisTwo += _mapBinSize;
+      } //end axis two loop
+      axisOne += _mapBinSize;
+    } //end axis one loop
+  } //end fillHistorgram
+
+  void BFieldPlotter::endJob() {
+
+  } // end endJob
+
+
+}  // end namespace mu2e
+
+using mu2e::BFieldPlotter;
+DEFINE_ART_MODULE(BFieldPlotter);
```

Comment:
You can write this as one line:

DEFINE_ART_MODULE(mu2e::BFieldPlotter);

In the early days of art you could not do it in one line due to limitations in CPP.  They are long gone.

Reply:
Clearly an inherited aspect, thank you!